app/notification.go

```diff
@@ -349,76 +349,156 @@ func sendNotificationEmail(post *model.Post, user *model.User, channel *model.Ch
-			"TimeZone": zone, "Month": month, "Day": day}))
+	var channelName = channel.DisplayName
+	if channel.Type == model.CHANNEL_GROUP {
+		channelName = translateFunc("api.templates.channel_name.group")
+	}
+	tm, month, day, _, zone := getFormattedPostTime(post, translateFunc)
 
-	go func() {
-		if err := utils.SendMail(user.Email, html.UnescapeString(subject), bodyPage.Render()); err != nil {
-			l4g.Error(utils.T("api.post.send_notifications_and_forget.send.error"), user.Email, err)
+	var bodyText string
+	var info string
+	if channel.Type == model.CHANNEL_DIRECT {
+		if emailNotificationContentsType == model.EMAIL_NOTIFICATION_CONTENTS_FULL {
+			bodyText = translateFunc("app.notification.body.intro.direct.full")
+			info = translateFunc("app.notification.body.text.direct.full",
+				map[string]interface{}{
+					"SenderName": senderName,
+					"Hour":       fmt.Sprintf("%02d", tm.Hour()),
+					"Minute":     fmt.Sprintf("%02d", tm.Minute()),
+					"TimeZone":   zone,
+					"Month":      month,
+					"Day":        day,
+				})
+		} else {
+			bodyText = translateFunc("app.notification.body.intro.direct.generic", map[string]interface{}{
+				"SenderName": senderName,
+			})
+			info = translateFunc("app.notification.body.text.direct.generic",
+				map[string]interface{}{
+					"Hour":     fmt.Sprintf("%02d", tm.Hour()),
+					"Minute":   fmt.Sprintf("%02d", tm.Minute()),
+					"TimeZone": zone,
+					"Month":    month,
+					"Day":      day,
+				})
+		}
+	} else {
+		if emailNotificationContentsType == model.EMAIL_NOTIFICATION_CONTENTS_FULL {
+			bodyText = translateFunc("app.notification.body.intro.notification.full")
+			info = translateFunc("app.notification.body.text.notification.full",
+				map[string]interface{}{
+					"ChannelName": channelName,
+					"SenderName":  senderName,
+					"Hour":        fmt.Sprintf("%02d", tm.Hour()),
+					"Minute":      fmt.Sprintf("%02d", tm.Minute()),
+					"TimeZone":    zone,
+					"Month":       month,
+					"Day":         day,
+				})
+		} else {
+			bodyText = translateFunc("app.notification.body.intro.notification.generic", map[string]interface{}{
+				"SenderName": senderName,
+			})
+			info = translateFunc("app.notification.body.text.notification.generic",
+				map[string]interface{}{
+					"Hour":     fmt.Sprintf("%02d", tm.Hour()),
+					"Minute":   fmt.Sprintf("%02d", tm.Minute()),
+					"TimeZone": zone,
+					"Month":    month,
+					"Day":      day,
+				})
 		}
-	}()
-
-	if einterfaces.GetMetricsInterface() != nil {
-		einterfaces.GetMetricsInterface().IncrementPostSentEmail()
 	}
 
-	return nil
+	bodyPage.Props["BodyText"] = bodyText
+	bodyPage.Html["Info"] = template.HTML(info)
+	bodyPage.Props["Button"] = translateFunc("api.templates.post_body.button")
+
+	return bodyPage.Render()
+}
+
+func getFormattedPostTime(post *model.Post, translateFunc i18n.TranslateFunc) (time.Time, string, string, string, string) {
```

Comment:
I think we might want a one-off type to use for the return value here so that
1. We can return hours and minutes without having 7 return values since we use both of those all over getNotificationEmailBody
2. We don't need to worry about what each return value is and whether this returns month/day/year/timezone, year/month/day/timezone, year/month/day/hour/second, etc